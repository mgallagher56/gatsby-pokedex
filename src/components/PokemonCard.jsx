import React from "react"
import { Link } from 'gatsby'
import { GatsbyImage } from 'gatsby-plugin-image'
import { makeStyles } from '@material-ui/core/styles'

import Button from '@material-ui/core/Button'
import Card from '@material-ui/core/Card'
import CardActions from '@material-ui/core/CardActions'
import CardContent from '@material-ui/core/CardContent'
import Typography from '@material-ui/core/Typography'
import { CardMedia } from '@material-ui/core'

// styles
const useStyles = makeStyles((theme) => ({
  card: {
    display: 'flex',
    flexDirection: 'column',
    height: '100%',
    transition: 'all .3s ease-out',
    transitionProperty: 'box-shadow, transform',
    '&:hover': {
      boxShadow: '0px 0px 10px rgba(0, 0, 0, 0.2)',
      transform: 'translate3d(0px, -2px, 0px)'
    },
    [theme.breakpoints.down('xs')]: {
      flexDirection: 'row',
      maxHeight: '100px'
    }
  },
  cardTitle: {
    textTransform: 'capitalize'
  },
  cardMedia: {
    [theme.breakpoints.down('xs')]: {
      width: '33%'
    }
  },
  cardContent: {
    flexGrow: 1,
    display: 'flex',
    alignItems: 'center'
  }
}))

const PokemonCard = (props) => {
  const {
    card,
    cardTitle,
    cardMedia,
    cardContent
  } = useStyles()

  return (
    <Link to={props.link}>
      <Card className={card}>
        {props.gatsbyImg
          ? <GatsbyImage
              objectFit='contain'
              className={cardMedia}
              image={props.gatsbyImg}
              title={props.title}
              alt={props.title}
            />
          : <CardMedia
              className={cardMedia}
              image={props.fallbackImg}
              title={props.title}
              alt={props.title}
            />}
        <CardContent className={cardContent}>
          <Typography variant='h5' component='h5' className={cardTitle}>
            {props.title}
          </Typography>
        </CardContent>
        <CardActions>
          <Button color='primary'>
            {props.btnText}
          </Button>
        </CardActions>
      </Card>
    </Link>
  )
}

export default PokemonCard
