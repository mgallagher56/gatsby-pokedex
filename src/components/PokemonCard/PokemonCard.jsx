import React from 'react'
import { Link } from 'gatsby'
import { GatsbyImage } from 'gatsby-plugin-image'
import { makeStyles } from '@material-ui/core/styles'

import Button from '@material-ui/core/Button'
import Card from '@material-ui/core/Card'
import CardActions from '@material-ui/core/CardActions'
import CardContent from '@material-ui/core/CardContent'
import Typography from '@material-ui/core/Typography'
import { CardMedia } from '@material-ui/core'

// styles
const useStyles = makeStyles((theme) => ({
  card: {
    display           : 'flex',
    flexDirection     : 'column',
    height            : '100%',
    transition        : 'all .3s ease-out',
    transitionProperty: 'box-shadow, transform',
    '&:hover' : {
      boxShadow: '0px 0px 10px rgba(0, 0, 0, 0.2)',
      transform: 'translate3d(0px, -2px, 0px)'
    },
    [theme.breakpoints.down('xs')]: {
      flexDirection: 'row',
      maxHeight    : '100px'
    }
  },
  cardTitle: {
    textTransform: 'capitalize'
  },
  cardMedia: {
    [theme.breakpoints.down('xs')]: {
      width: '33%'
    }
  },
  cardContent: {
    flexGrow: 1,
    display: 'flex',
    alignItems: 'left',
    [theme.breakpoints.down('xs')]: {
      alignItems: 'center'
    }
  }
}))

/**
 * PokemonCard renders a linked card component
 *
 * Props:
 *
 * link - adds a link to whole card component
 *
 * title - adds title to card
 *
 * description - add a paragraph to provide further detail
 *
 * gatsbyImg - adds gatbsy image object for responsive images
 *
 * fallbackImg - adds a fallback image to render if gatsby image object fails
 *
 * btnText - adds button to card and populates with text
 */
const PokemonCard = (props) => {
  // destructure styles for increased legibility
  const {
    card,
    cardTitle,
    cardMedia,
    cardContent,
    cardActions
  } = useStyles()

  // markup
  return (
    <div>
      {props.link
        ? <Link to={props.link || '/'} data-testid='pokemon-card-link'>
          <Card className={`${card} pokemon-card`} src={props.gatsbyImg}>
            {props.gatsbyImg
              ? <GatsbyImage
                  component='img'
                  objectFit='contain'
                  className={cardMedia}
                  image={props.gatsbyImg}
                  title={props.title}
                  alt={props.title}
                />
              : <CardMedia
                  className={cardMedia}
                  image={props.fallbackImg}
                  title={props.title}
                  alt={props.title}
                  component='img'
                />}
            <CardContent className={`${cardContent} card-content`}>
              {props.title
                ? <Typography data-testid='pokemon-card-title' variant='h5' component='h2' className={cardTitle}>
                  {props.title}
                </Typography>
                : null}
              {props.description
                ? <Typography data-testid='pokemon-card-description' variant='body1' component='p'>
                  {props.description}
                </Typography>
                : null}
            </CardContent>
            {props.btnText
              ? <CardActions className={cardActions}>
                <Button data-testid='pokemon-card-button' color='primary'>{props.btnText}</Button>
              </CardActions>
              : null}
          </Card>
        </Link>
        : null}
    </div>
  )
}

export default PokemonCard
