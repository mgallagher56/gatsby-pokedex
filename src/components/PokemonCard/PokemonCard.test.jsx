import React from 'react'
import { render, cleanup } from '@testing-library/react'
import PokemonCard from './PokemonCard'

describe('PokemonCard Component', () => {
  afterEach(cleanup)

  it('renders PokemonCard Component', () => {
    render(<PokemonCard link='/' />)
  })

  it('should take a snapshot', () => {
    const { asFragment } = render(<PokemonCard link='/' />)
    expect(asFragment(<PokemonCard link='/' />)).toMatchSnapshot()
  })

  it('should have Hello world title', () => {
    const { getByTestId } = render(<PokemonCard link='/' title='Hello world' />)
    expect(getByTestId('pokemon-card-title')).toMatchInlineSnapshot(`
    <h2
      class="MuiTypography-root makeStyles-cardTitle-10 MuiTypography-h5"
      data-testid="pokemon-card-title"
    >
      Hello world
    </h2>
    `)
  })

  it('should have a title', () => {
    const { queryByTestId } = render(<PokemonCard link='/' title='title text' />)
    expect(queryByTestId('pokemon-card-title')).toBeTruthy()
  })

  it('should not have a title', () => {
    const { queryByTestId } = render(<PokemonCard link='/' />)
    expect(queryByTestId('pokemon-card-title')).toBeNull()
  })

  it('should have a description', () => {
    const { queryByTestId } = render(<PokemonCard link='/' description='lorem ipsum' />)
    expect(queryByTestId('pokemon-card-description')).toBeTruthy()
  })

  it('should not have a description', () => {
    const { queryByTestId } = render(<PokemonCard link='/' />)
    expect(queryByTestId('pokemon-card-description')).toBeNull()
  })

  it('should have a button', () => {
    const { queryByTestId } = render(<PokemonCard link='/' btnText='Go back' />)
    expect(queryByTestId('pokemon-card-button')).toBeTruthy()
  })

  it('should not have a button', () => {
    const { queryByTestId } = render(<PokemonCard link='/' />)
    expect(queryByTestId('pokemon-card-button')).toBeNull()
  })

  it('should be a link', () => {
    const { getByText } = render(<PokemonCard link='/pokemon' title='test' />)
    expect(document.querySelector('a').getAttribute('href')).toBe('/pokemon')
  })

  it('should not exist', () => {
    const { queryByTestId } = render(<PokemonCard />)
    expect(queryByTestId('pokemon-card-link')).toBeNull()
  })
})
