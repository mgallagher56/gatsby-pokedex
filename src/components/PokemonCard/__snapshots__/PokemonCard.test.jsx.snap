// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonCard Component should take a snapshot 1`] = `
<DocumentFragment>
  <div>
    <a
      data-testid="pokemon-card-link"
      href="/"
    >
      <div
        class="MuiPaper-root MuiCard-root makeStyles-card-5 pokemon-card MuiPaper-elevation1 MuiPaper-rounded"
      >
        <img
          class="MuiCardMedia-root makeStyles-cardMedia-7 MuiCardMedia-media MuiCardMedia-img"
        />
        <div
          class="MuiCardContent-root makeStyles-cardContent-8 card-content"
        />
      </div>
    </a>
  </div>
</DocumentFragment>
`;
